### ISSUE TITLE
DeviceIoControlAsync incorrent pointer offset of input/output buffer

### ISSUE THREAD

**Issue by GreatFireWall** [dahall/Vanara#22]
Hi, forgive me for my broken English.

**Describe the bug**
`DeviceIoControlAsync` incorrent pointer offset of input/output buffer.

**What code is involved**
`Vanara.PInvoke.Kernel32.DeviceIoControlAsync` at line `133` in file `Vanara\PInvoke\Kernel32\IoApiSet.Threading.cs`

`Vanara.PInvoke.Kernel32.Pack` at line `915` in file `Vanara\PInvoke\Kernel32\IoApiSet.cs`

`Vanara.PInvoke.Kernel32.BeginDeviceIoControl` at line `887` in file `Vanara\PInvoke\Kernel32\IoApiSet.cs`

**Expected behavior**
Rather than the `buf` that contains struct lengths wrote in
`Vanara\PInvoke\Kernel32\IoApiSet.cs` at line `915`:
```cs
private static byte[] Pack<TIn, TOut>(TIn? inVal, TOut? outVal) where TIn : struct where TOut : struct
{
   using (var ms = new MemoryStream())
   using (var wtr = new BinaryWriter(ms))
   {
      wtr.Write(inVal.HasValue ? Marshal.SizeOf(typeof(TIn)) : 0);
      wtr.Write(outVal.HasValue ? Marshal.SizeOf(typeof(TOut)) : 0);
      if (inVal.HasValue) wtr.Write(inVal.Value);
      if (outVal.HasValue) wtr.Write(outVal.Value);
      return ms.ToArray();
   }
}
```
, the `DeviceIoControlAsync` should pass a `buf` that skips the lengths of `TIn` and `TOut`

**Screenshots**

![image](https://user-images.githubusercontent.com/4510984/48663560-dd30f000-eacc-11e8-8a48-1a303ad65c44.png)

![image](https://user-images.githubusercontent.com/4510984/48663595-6fd18f00-eacd-11e8-9ada-62ab08ad7866.png)

**Comment by dahall** [comment]
Does this look correct?
```cs
private static unsafe IAsyncResult BeginDeviceIoControl<TIn, TOut>(HFILE hDevice, uint dwIoControlCode, byte[] buffer, AsyncCallback userCallback, object userState) where TIn : struct where TOut : struct
{
   var ar = OverlappedAsync.SetupOverlappedFunction(hDevice, userCallback, buffer);
   var prefix = Marshal.SizeOf(typeof(int)) * 2;
   var inSz = Marshal.SizeOf(typeof(TIn));
   fixed (byte* pIn = &buffer[prefix], pOut = &buffer[prefix + inSz])
   {
      var ret = DeviceIoControl(hDevice, dwIoControlCode, pIn, (uint)inSz, pOut, (uint)Marshal.SizeOf(typeof(TOut)), out var bRet, ar.Overlapped);
      return OverlappedAsync.EvaluateOverlappedFunction(ar, ret);
   }
}
```

**Comment by GreatFireWall** [comment]
it's ok, could you update it to nuget as soon as possible ? thank you.

**Comment by dahall** [comment]
I'm in the process of testing major changes to the libraries, but expect that testing to take the next 2 weeks. I will release version 2.0 at that time and include this fix.

**Comment by GreatFireWall** [comment]
ok, thanks.

**Comment by dahall** [comment]
Thank you for all your work on this! I will update the code with your fixes and post it later today.

**Comment by dahall** [comment]
Fixed in 2.0 release today
